qurro/qarcoal.py

```diff
@@ -0,0 +1,112 @@
+    num_string in numerator over denom_string in denominator.
+
+    Parameters:
+    -----------
+        table: biom file with which to calculate log ratios
+        taxonomy: Qiime2 compliant feature taxonomy metadata file
+        num_string: numerator string to search for in taxonomy
+        denom_string: denominator string to search for in taxonomy
+        sample_to_use: Q2 Metadata file with samples to use (optional)
+        allow_shared_features: bool denoting whether to raise error
+            if features are shared between numerator and denominator
+            (default: False)
+
+    Returns:
+    --------
+        comparison_df: pd DataFrame in the form:
+
+            Sample-ID    Num_Sum    Denom_Sum   log_ratio
+                   S1          7           15   -0.762140
+    """
+
+    # biom table is features x samples
+    if samples_to_use is not None:
+        samp = set(samples_to_use.to_dataframe().index)
+        feat_table = table.filter(samp, axis="sample", inplace=False)
+        feat_table = feat_table.to_dataframe()
+    else:
+        feat_table = table.to_dataframe()
+
+    taxonomy_df = taxonomy.to_dataframe()
+
+    # taxonomy is features x [Taxon, Confidence]
+    taxonomy_df = taxonomy_df.loc[feat_table.index]
+    tax_num_df = taxonomy_df[taxonomy_df["Taxon"].str.contains(num_string)]
+    tax_denom_df = taxonomy_df[taxonomy_df["Taxon"].str.contains(denom_string)]
+
+    if tax_num_df.shape[0] == 0:
+        if tax_denom_df.shape[0] == 0:
+            raise ValueError("neither feature found!")
+        else:
+            raise ValueError("numerator not found!")
+    elif tax_denom_df.shape[0] == 0:
+        raise ValueError("denominator not found!")
+    else:
+        pass
+
+    # drop columns (samples) in which no feature w/ string is present
+    tax_num_df.drop(columns=["Taxon", "Confidence"], inplace=True)
+    tax_denom_df.drop(columns=["Taxon", "Confidence"], inplace=True)
+
+    tax_num_joined_df = tax_num_df.join(feat_table)
+    tax_num_joined_df = tax_num_joined_df.loc[
+        :, (tax_num_joined_df != 0).any(axis=0)
+    ]
+    tax_denom_joined_df = tax_denom_df.join(feat_table)
+    tax_denom_joined_df = tax_denom_joined_df.loc[
+        :, (tax_denom_joined_df != 0).any(axis=0)
+    ]
+
+    # keep only intersection of samples in which each feat string
+    # is present
+    samp_to_keep = set(tax_num_joined_df.columns).intersection(
+        set(tax_denom_joined_df.columns)
+    )
+    tax_num_joined_df = tax_num_joined_df[samp_to_keep]
+    tax_denom_joined_df = tax_denom_joined_df[samp_to_keep]
+
+    tax_num_sample_sum = tax_num_joined_df.sum(axis=0)
+    tax_denom_sample_sum = tax_denom_joined_df.sum(axis=0)
+
+    # if shared features are disallowed, check to make sure they don't occur
+    # if allowed, can skip this step at user's risk
+    if not allow_shared_features:
+        shared_features = set(tax_num_df.index) & set(tax_denom_df.index)
+        if shared_features:
```

Comment:
I'm pretty sure this code will work as written, but implicitly converting other stuff to booleans in `if` statements always makes me anxious. I think it'd be safer (or at least clearer) to explicitly write out `if len(shared_features) > 0:` here.

Reply:
Curious to hear your thoughts on this from a stylistic point. The `if shared_features` conditional is checking the "Falsiness" of the set, so my understanding is there's not a conversion. I've seen people say that this way is "more Pythonic" but I'm not at all tied to it.

https://realpython.com/python-pep8/

(Ctrl-F "falsy")

Reply:
Also I can't find this in the actual pep8 documentation so... 🤷‍♂

Reply:
[Found it](https://www.python.org/dev/peps/pep-0008/#programming-recommendations)
>For sequences, (strings, lists, tuples), use the fact that empty sequences are false.

That being said I 100% ignore pep on occasion so it's up to you.

Reply:
Huh. Apparently I've been violating PEP8 for forever :P I see your point re: the conversion -- that makes sense, "conversion" is probably the wrong word for that situation.

I'm cool with keeping this as is -- either way is ok, but I think following PEP8 here makes sense. Thank you for pushing back on this, I appreciate it :)